File: app/src/main/java/com/rubyhuntersky/peregrine/ui/BaseActivity.java
package com.rubyhuntersky.peregrine.ui;

import android.app.AlertDialog;
import android.app.FragmentManager;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.util.Pair;

import com.rubyhuntersky.peregrine.AccountAssets;
import com.rubyhuntersky.peregrine.AllAccounts;
import com.rubyhuntersky.peregrine.EtradeAccount;
import com.rubyhuntersky.peregrine.EtradeApi;
import com.rubyhuntersky.peregrine.PartitionList;
import com.rubyhuntersky.peregrine.PortfolioAssets;
import com.rubyhuntersky.peregrine.R;
import com.rubyhuntersky.peregrine.Storage;
import com.rubyhuntersky.peregrine.exception.NotStoredException;
import com.rubyhuntersky.peregrine.exception.ProductionStorage;
import com.rubyhuntersky.peregrine.oauth.OauthToken;
import com.rubyhuntersky.peregrine.oauth.OauthVerifier;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.DateFormat;
import java.util.Date;
import java.util.List;

import rx.Observable;
import rx.Subscription;
import rx.functions.Action0;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.functions.Func2;
import rx.subjects.BehaviorSubject;
import rx.subscriptions.Subscriptions;

import static com.rubyhuntersky.peregrine.ui.oauth.OauthUi.promptForVerifier;

/**
 * @author wehjin
 * @since 11/21/15.
 */

public class BaseActivity extends AppCompatActivity {

    protected final String TAG = this.getClass().getSimpleName();
    protected Action1<Throwable> errorAction = new Action1<Throwable>() {
        @Override
        public void call(Throwable throwable) {
            final String title = "ERROR";
            Log.e(TAG, title, throwable);
            alertError(title, throwable);
        }
    };
    private Subscription refreshSubscription = Subscriptions.empty();
    private EtradeApi etradeApi;
    private Storage storage;
    private BehaviorSubject<PartitionList> partitionListStream;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        etradeApi = new EtradeApi(this);
        storage = new ProductionStorage(this, etradeApi.oauthAppToken);

        try {
            final InputStream inputStream = getResources().openRawResource(R.raw.starting_partitions);
            partitionListStream = BehaviorSubject.create(new PartitionList(new JSONObject(getString(inputStream))));
        } catch (JSONException | IOException e) {
            Log.e(TAG, "onCreate", e);
            throw new RuntimeException(e);
        }
    }

    @NonNull
    private String getString(InputStream inputStream) throws IOException {
        StringBuilder stringBuilder = new StringBuilder();
        BufferedReader br = new BufferedReader(new InputStreamReader(inputStream));
        for (String s = br.readLine(); s != null; s = br.readLine())
            stringBuilder.append(s);
        for (int c = br.read(); c != -1; c = br.read())
            stringBuilder.append((char) c);
        return stringBuilder.toString();
    }

    public Observable<PartitionList> getPartitionListStream() {
        return partitionListStream;
    }

    public Observable<AllAccounts> getAllAccountsStream() {
        return storage.streamAccountsList();
    }


    public Observable<List<AccountAssets>> getAccountAssetsListStream() {
        return storage.streamAccountAssetsList();
    }

    public Observable<PortfolioAssets> getPortfolioAssetsStream() {
        return storage.streamAccountAssetsList().map(new Func1<List<AccountAssets>, PortfolioAssets>() {
            @Override
            public PortfolioAssets call(List<AccountAssets> accountAssetsList) {
                return new PortfolioAssets(accountAssetsList);
            }
        });
    }

    public void refresh() {
        refreshSubscription.unsubscribe();
        refreshSubscription =
              fetchAndStoreAccountsList().flatMap(new Func1<AllAccounts, Observable<List<AccountAssets>>>() {
                  @Override
                  public Observable<List<AccountAssets>> call(AllAccounts allAccounts) {
                      return fetchAndStoreAccountAssetsList(allAccounts);
                  }
              }).subscribe(new Action1<List<AccountAssets>>() {
                  @Override
                  public void call(List<AccountAssets> accountAssetList) {
                      logDebug("Refresh completed");
                  }
              }, new Action1<Throwable>() {
                  @Override
                  public void call(Throwable throwable) {
                      alertError("Refresh error", throwable);
                  }
              });
    }

    public EtradeApi getEtradeApi() {
        return etradeApi;
    }

    public Storage getStorage() {
        return storage;
    }

    @Override
    public void onBackPressed() {
        final FragmentManager fragmentManager = getFragmentManager();
        if (fragmentManager.getBackStackEntryCount() != 0) {
            fragmentManager.popBackStack();
        } else {
            super.onBackPressed();
        }
    }

    @NonNull
    private Observable<List<AccountAssets>> fetchAndStoreAccountAssetsList(AllAccounts allAccounts) {
        return (allAccounts == null
                ? Observable.just((List<AccountAssets>) null)
                : fetchAccountAssets(allAccounts).toList()).doOnNext(new Action1<List<AccountAssets>>() {
            @Override
            public void call(List<AccountAssets> accountAssetsList) {
                getStorage().writeAccountAssetsList(accountAssetsList);
            }
        });
    }

    @NonNull
    private Observable<AccountAssets> fetchAccountAssets(AllAccounts allAccounts) {
        return Observable.from(allAccounts.accounts)
                         .flatMap(new Func1<EtradeAccount, Observable<Pair<JSONObject, JSONObject>>>() {
                             @Override
                             public Observable<Pair<JSONObject, JSONObject>> call(EtradeAccount etradeAccount) {
                                 return fetchAccountDetails(etradeAccount);
                             }
                         })
                         .map(new Func1<Pair<JSONObject, JSONObject>, AccountAssets>() {
                             @Override
                             public AccountAssets call(Pair<JSONObject, JSONObject> jsonPair) {
                                 try {
                                     return new AccountAssets(EtradeApi.addBalanceToPositions(jsonPair.first, jsonPair.second));
                                 } catch (JSONException e) {
                                     throw new RuntimeException(e);
                                 }
                             }
                         });
    }


    protected Observable<AllAccounts> fetchAndStoreAccountsList() {
        final Func1<OauthToken, Observable<List<EtradeAccount>>> accessTokenToAccountList =
              new Func1<OauthToken, Observable<List<EtradeAccount>>>() {
                  @Override
                  public Observable<List<EtradeAccount>> call(OauthToken oauthToken) {
                      return getEtradeApi().fetchAccountList(oauthToken);
                  }
              };
        return getOauthAccessToken().flatMap(accessTokenToAccountList)
                                    .onErrorResumeNext(new Func1<Throwable, Observable<? extends List<EtradeAccount>>>() {
                                        @Override
                                        public Observable<? extends List<EtradeAccount>> call(Throwable throwable) {
                                            if (throwable instanceof EtradeApi.NotAuthorizedException) {
                                                return renewOauthAccessToken().flatMap(accessTokenToAccountList);
                                            } else {
                                                return Observable.error(throwable);
                                            }
                                        }
                                    })
                                    .onErrorResumeNext(new Func1<Throwable, Observable<? extends List<EtradeAccount>>>() {
                                        @Override
                                        public Observable<? extends List<EtradeAccount>> call(Throwable throwable) {
                                            if (throwable instanceof EtradeApi.NotAuthorizedException) {
                                                return fetchOauthAccessToken().flatMap(accessTokenToAccountList)
                                                                              .doOnSubscribe(new Action0() {
                                                                                  @Override
                                                                                  public void call() {
                                                                                      getStorage().eraseOauthAccessToken();
                                                                                  }
                                                                              });
                                            } else {
                                                return Observable.error(throwable);
                                            }
                                        }
                                    })
                                    .map(new Func1<List<EtradeAccount>, AllAccounts>() {
                                        @Override
                                        public AllAccounts call(List<EtradeAccount> etradeAccounts) {
                                            return new AllAccounts(etradeAccounts, new Date());
                                        }
                                    })
                                    .doOnNext(new Action1<AllAccounts>() {
                                        @Override
                                        public void call(AllAccounts allAccounts) {
                                            getStorage().writeAccountList(allAccounts);
                                        }
                                    });
    }

    private Observable<Pair<JSONObject, JSONObject>> fetchAccountDetails(final EtradeAccount etradeAccount) {
        return getStorage().readOauthAccessToken()
                           .flatMap(new Func1<OauthToken, Observable<Pair<JSONObject, JSONObject>>>() {
                               @Override
                               public Observable<Pair<JSONObject, JSONObject>> call(OauthToken oauthToken) {
                                   return fetchAccountPositionsAndBalanceResponses(oauthToken, etradeAccount);
                               }
                           });
    }

    @NonNull
    private Observable<Pair<JSONObject, JSONObject>> fetchAccountPositionsAndBalanceResponses(OauthToken oauthToken, EtradeAccount etradeAccount) {
        final Func2<JSONObject, JSONObject, Pair<JSONObject, JSONObject>> toPair =
              new Func2<JSONObject, JSONObject, Pair<JSONObject, JSONObject>>() {
                  @Override
                  public Pair<JSONObject, JSONObject> call(JSONObject positions, JSONObject balance) {
                      return new Pair<>(positions, balance);
                  }
              };
        return Observable.zip(fetchAccountPositionsResponse(oauthToken, etradeAccount), fetchAccountBalanceResponse(oauthToken, etradeAccount), toPair);
    }

    @NonNull
    private Observable<JSONObject> fetchAccountPositionsResponse(OauthToken oauthToken, final EtradeAccount etradeAccount) {
        return getEtradeApi().fetchAccountPositionsResponse(etradeAccount.accountId, oauthToken)
                             .retry(1)
                             .map(new Func1<JSONObject, JSONObject>() {
                                 @Override
                                 public JSONObject call(JSONObject jsonObject) {
                                     try {
                                         jsonObject.putOpt("accountDescription", etradeAccount.description);
                                         jsonObject.putOpt("requestAccountId", etradeAccount.accountId);
                                         jsonObject.putOpt("responseArrivalTime", DateFormat.getInstance()
                                                                                            .format(new Date()));
                                     } catch (JSONException e) {
                                         throw new RuntimeException(e);
                                     }
                                     return jsonObject;
                                 }
                             });
    }

    @NonNull
    private Observable<JSONObject> fetchAccountBalanceResponse(OauthToken oauthToken, final EtradeAccount etradeAccount) {
        return getEtradeApi().fetchAccountBalanceResponse(etradeAccount.accountId, oauthToken)
                             .retry(1)
                             .map(new Func1<JSONObject, JSONObject>() {
                                 @Override
                                 public JSONObject call(JSONObject jsonObject) {
                                     try {
                                         jsonObject.putOpt("accountDescription", etradeAccount.description);
                                         jsonObject.putOpt("requestAccountId", etradeAccount.accountId);
                                         jsonObject.putOpt("responseArrivalTime", DateFormat.getInstance()
                                                                                            .format(new Date()));
                                     } catch (JSONException e) {
                                         throw new RuntimeException(e);
                                     }
                                     return jsonObject;
                                 }
                             })
                             .doOnError(new Action1<Throwable>() {
                                 @Override
                                 public void call(Throwable throwable) {
                                     Log.e(TAG, " v error, account: " + etradeAccount);
                                 }
                             });
    }


    private Observable<OauthToken> getOauthAccessToken() {
        return getStorage().readOauthAccessToken()
                           .onErrorResumeNext(new Func1<Throwable, Observable<? extends OauthToken>>() {
                               @Override
                               public Observable<? extends OauthToken> call(Throwable throwable) {
                                   if (throwable instanceof NotStoredException) {
                                       return fetchOauthAccessToken();
                                   } else {
                                       return Observable.error(throwable);
                                   }
                               }
                           });
    }

    private Observable<OauthToken> renewOauthAccessToken() {
        return getOauthAccessToken().flatMap(new Func1<OauthToken, Observable<OauthToken>>() {
            @Override
            public Observable<OauthToken> call(OauthToken oauthToken) {
                return getEtradeApi().renewOauthAccessToken(oauthToken);
            }
        });
    }

    private Observable<OauthToken> fetchOauthAccessToken() {
        return getEtradeApi().fetchOauthRequestToken().flatMap(new Func1<OauthToken, Observable<OauthVerifier>>() {
            @Override
            public Observable<OauthVerifier> call(OauthToken requestToken) {
                return promptForVerifier(BaseActivity.this, requestToken);
            }
        }).flatMap(new Func1<OauthVerifier, Observable<OauthToken>>() {
            @Override
            public Observable<OauthToken> call(OauthVerifier verifier) {
                return getEtradeApi().fetchOauthAccessToken(verifier);
            }
        }).doOnNext(new Action1<OauthToken>() {
            @Override
            public void call(OauthToken oauthToken) {
                getStorage().writeOauthAccessToken(oauthToken);
            }
        });
    }

    protected void alertError(String title, Throwable throwable) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title)
               .setMessage(throwable.toString())
               .setPositiveButton("Close", new DialogInterface.OnClickListener() {
                   @Override
                   public void onClick(DialogInterface dialog, int which) {
                       dialog.dismiss();
                   }
               })
               .show();
        logError(title, throwable);
    }

    protected void logDebug(String message) {
        Log.d(TAG, message);
    }

    protected void logError(String message, Throwable throwable) {
        Log.e(TAG, message + "\n" + throwable.getLocalizedMessage());
    }
}

File: app/src/main/java/com/rubyhuntersky/peregrine/ui/BaseActivity.kt
package com.rubyhuntersky.peregrine.ui

import android.app.AlertDialog
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.util.Pair
import com.rubyhuntersky.peregrine.R
import com.rubyhuntersky.peregrine.exception.NotStoredException
import com.rubyhuntersky.peregrine.exception.ProductionStorage
import com.rubyhuntersky.peregrine.lib.oauth.model.OauthToken
import com.rubyhuntersky.peregrine.lib.oauth.ui.OauthUi.promptForVerifier
import com.rubyhuntersky.peregrine.model.*
import org.json.JSONException
import org.json.JSONObject
import rx.Observable
import rx.Subscription
import rx.functions.Action1
import rx.functions.Func1
import rx.subjects.BehaviorSubject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.text.DateFormat
import java.util.*
import java.util.concurrent.TimeUnit

abstract class BaseActivity : AppCompatActivity() {
    var refreshSubscription: Subscription? = null
    lateinit var etradeApi: EtradeApi
    lateinit var storage: Storage
    lateinit var partitionListStream: BehaviorSubject<PartitionList>
    val tag: String = this.javaClass.simpleName
    val errorAction: Action1<Throwable> = Action1 { throwable ->
        val title = "ERROR"
        Log.e(tag, title, throwable)
        alertError(title, throwable)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        etradeApi = EtradeApi(this)
        storage = ProductionStorage(this, etradeApi.oauthAppToken)
        try {
            val inputStream = resources.openRawResource(R.raw.starting_partitions)
            partitionListStream = BehaviorSubject.create(PartitionList(JSONObject(getString(inputStream))))
        } catch (e: JSONException) {
            Log.e(tag, "onCreate", e)
            throw RuntimeException(e)
        } catch (e: IOException) {
            Log.e(tag, "onCreate", e)
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    private fun getString(inputStream: InputStream): String {
        val stringBuilder = StringBuilder()
        val br = BufferedReader(InputStreamReader(inputStream))
        var s: String? = br.readLine()
        while (s != null) {
            stringBuilder.append(s)
            s = br.readLine()
        }
        var c = br.read()
        while (c != -1) {
            stringBuilder.append(c.toChar())
            c = br.read()
        }
        return stringBuilder.toString()
    }

    fun getPartitionListStream(): Observable<PartitionList> {
        return partitionListStream
    }

    val allAccountsStream: Observable<AllAccounts> get() = storage.streamAccountsList()
    val accountAssetsListStream: Observable<List<AccountAssets>> get() = storage.streamAccountAssetsList()
    val portfolioAssetsStream: Observable<PortfolioAssets> get() = storage.streamAccountAssetsList().map(::PortfolioAssets)

    fun refresh() {
        refreshSubscription?.unsubscribe()
        refreshSubscription = fetchAndStoreAccountsList().flatMap { allAccounts ->
            fetchAndStoreAccountAssetsList(allAccounts)
        }.subscribe({ logDebug("Refresh completed") }, { alertError("Refresh error", it) })
    }

    override fun onBackPressed() {
        val fragmentManager = fragmentManager
        if (fragmentManager.backStackEntryCount != 0) {
            fragmentManager.popBackStack()
        } else {
            super.onBackPressed()
        }
    }

    private fun fetchAndStoreAccountAssetsList(allAccounts: AllAccounts): Observable<List<AccountAssets>> = fetchAccountAssets(allAccounts).toList().doOnNext { storage.writeAccountAssetsList(it) }

    private fun fetchAccountAssets(allAccounts: AllAccounts): Observable<AccountAssets> {
        return Observable.from(allAccounts.accounts).flatMap(object : Func1<EtradeAccount, Observable<Pair<JSONObject, JSONObject>>> {
            internal var count = 0

            override fun call(etradeAccount: EtradeAccount): Observable<Pair<JSONObject, JSONObject>> {
                return fetchAccountDetails(etradeAccount).delaySubscription((200 * count++).toLong(), TimeUnit.MILLISECONDS)
            }
        }).map(Func1<Pair<JSONObject, JSONObject>, com.rubyhuntersky.peregrine.model.AccountAssets> { jsonPair ->
            try {
                return@Func1 AccountAssets(EtradeApi.addBalanceToPositions(jsonPair.first, jsonPair.second))
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }
        })
    }

    protected fun fetchAndStoreAccountsList(): Observable<AllAccounts> {
        return oauthAccessToken.flatMap({ it.toEtradeAccounts })
                .onErrorResumeNext { throwable ->
                    when (throwable) {
                        is EtradeApi.NotAuthorizedException -> renewOauthAccessToken().flatMap({ it.toEtradeAccounts })
                        else -> Observable.error(throwable)
                    }
                }
                .onErrorResumeNext { throwable ->
                    when (throwable) {
                        is EtradeApi.NotAuthorizedException -> {
                            storage.eraseOauthAccessToken()
                            fetchOauthAccessToken().flatMap({ it.toEtradeAccounts })
                        }
                        else -> Observable.error(throwable)
                    }
                }
                .map { AllAccounts(it, Date()) }
                .doOnNext { storage.writeAccountList(it) }
    }

    private fun fetchAccountDetails(etradeAccount: EtradeAccount): Observable<Pair<JSONObject, JSONObject>> {
        return storage.readOauthAccessToken().flatMap { oauthToken ->
            fetchAccountPositionsAndBalanceResponses(oauthToken, etradeAccount)
        }
    }

    private fun fetchAccountPositionsAndBalanceResponses(oauthToken: OauthToken, etradeAccount: EtradeAccount): Observable<Pair<JSONObject, JSONObject>> {
        return Observable.zip(
                fetchAccountPositionsResponse(oauthToken, etradeAccount),
                fetchAccountBalanceResponse(oauthToken, etradeAccount),
                { positions, balance -> Pair(positions, balance) }
        )
    }

    private fun fetchAccountPositionsResponse(oauthToken: OauthToken, etradeAccount: EtradeAccount): Observable<JSONObject> {
        return etradeApi.fetchAccountPositionsResponse(etradeAccount.accountId, oauthToken).retry(1).map { jsonObject ->
            try {
                jsonObject.putOpt("accountDescription", etradeAccount.description)
                jsonObject.putOpt("requestAccountId", etradeAccount.accountId)
                jsonObject.putOpt("responseArrivalTime", DateFormat.getInstance().format(Date()))
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }

            jsonObject
        }
    }

    private fun fetchAccountBalanceResponse(oauthToken: OauthToken, etradeAccount: EtradeAccount): Observable<JSONObject> {
        return etradeApi.fetchAccountBalanceResponse(etradeAccount.accountId, oauthToken).retry(1).map { jsonObject ->
            try {
                jsonObject.putOpt("accountDescription", etradeAccount.description)
                jsonObject.putOpt("requestAccountId", etradeAccount.accountId)
                jsonObject.putOpt("responseArrivalTime", DateFormat.getInstance().format(Date()))
            } catch (e: JSONException) {
                throw RuntimeException(e)
            }
            jsonObject
        }.doOnError { Log.e(tag, " v error, account: " + etradeAccount) }
    }


    private val oauthAccessToken: Observable<OauthToken>
        get() = storage.readOauthAccessToken().onErrorResumeNext { throwable ->
            when (throwable) {
                is NotStoredException -> fetchOauthAccessToken()
                else -> Observable.error(throwable)
            }
        }

    private fun renewOauthAccessToken(): Observable<OauthToken> = oauthAccessToken
            .flatMap { etradeApi.renewOauthAccessToken(it) }

    private fun fetchOauthAccessToken(): Observable<OauthToken> = etradeApi.fetchOauthRequestToken()
            .flatMap { requestToken -> promptForVerifier(this@BaseActivity, requestToken) }
            .flatMap { verifier -> etradeApi.fetchOauthAccessToken(verifier) }
            .doOnNext { oauthToken -> storage.writeOauthAccessToken(oauthToken) }

    private val OauthToken.toEtradeAccounts: Observable<List<EtradeAccount>> get() = etradeApi.fetchAccountList(this)

    protected fun alertError(title: String, throwable: Throwable) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(title).setMessage(throwable.toString()).setPositiveButton("Close") { dialog, which -> dialog.dismiss() }.show()
        logError(title, throwable)
    }

    protected fun logDebug(message: String) {
        Log.d(tag, message)
    }

    protected fun logError(message: String, throwable: Throwable) {
        Log.e(tag, "$message\n${throwable.message}")
    }
}
